options.shouldConsumeRetry

timkelty opened the pull request:
Allows an error to not count against the retry total.

An alternative to https://github.com/sindresorhus/p-retry/pull/94, suggested [here](https://github.com/sindresorhus/p-retry/pull/94#issuecomment-3186009471).

timkelty committed:
Implement shouldSkip

sindresorhus commented:
AI review:

---

## API design

* `shouldSkip(error)` is asymmetric with the rest of the API. Both `onFailedAttempt` and `shouldRetry` receive a **context** object since v7; `shouldSkip` should too. Otherwise users cannot decide based on `attemptNumber`, `retriesLeft`, etc. ([[GitHub](https://github.com/sindresorhus/p-retry)][1])
* Name is ambiguous. “Skip what?” Consider `shouldNotConsumeRetry(context)` or `shouldIgnoreForRetryBudget(context)`. If you keep `shouldSkip`, the docs must explicitly say “skip counting this failure toward `retries`”.
* It should support async like `shouldRetry`. Make signature `boolean | Promise<boolean>` for parity. ([[GitHub](https://github.com/sindresorhus/p-retry)][1])
* Pass the **normalized** error (the same one users see in `context.error`). Today you pass the raw thrown value to `shouldSkip`, which can be non-`Error` while the d.ts says `Error`.

## Ordering / correctness

Right now you:

1. build context,
2. run `onAttemptFailure(context)` which performs the exhaustion checks,
3. **then** call `shouldSkip(error)` and decrement `attemptNumber`.

Problems:

* With `retries: 0`, a “skippable” error on attempt 1 will still be rejected because `retriesLeft <= 0` triggers **before** the decrement. That violates “does not count against the retry limit.”
* Backoff grows even for skipped errors because the delay is computed before the decrement. If the intent is “doesn’t consume attempts,” it should also not inflate exponential backoff for the next *real* attempt.

Fix: decide `skip` **before** exhaustion checks and backoff.

```diff
- const context = createRetryContext(error, attemptNumber, options);
- await onAttemptFailure(context, options, startTime, maxRetryTime);
- if (options.shouldSkip(error)) {
- 	attemptNumber = Math.max(0, attemptNumber - 1);
- }
+ const skip = await options.shouldSkip?.({error, attemptNumber, retriesLeft: options.retries - (attemptNumber - 1)});
+ const effectiveAttempt = skip ? Math.max(1, attemptNumber - 1) : attemptNumber;
+ const context = createRetryContext(error, effectiveAttempt, options);
+ await onAttemptFailure(context, options, startTime, maxRetryTime);
+ if (skip) {
+ 	attemptNumber--;
+ }
```

Alternatively, change `onAttemptFailure` to accept `{skip}` and internally treat `retriesLeft + 1` when `skip` is true, and calculate delay from `effectiveAttempt`.

## Types

* `index.d.ts` should be:

```ts
readonly shouldSkip?: (context: RetryContext) => boolean | Promise<boolean>;
```

not `(error: Error) => boolean`. Non-`Error` throws are valid JS; your current d.ts lies.

* Add type tests in `index.test-d.ts` to cover the option.

## Docs

* Document whether skipped errors **affect backoff**. Today they do (because of ordering). Either change the code as above or say: “Skipped errors still incur the normal backoff delay.” Put this right under the option.
* Mention interaction with `AbortError` and non-network `TypeError`: they are thrown before any skipping logic, so they **cannot** be skipped. Add a note (and ideally a test) to prevent surprises. ([[GitHub](https://github.com/sindresorhus/p-retry)][1])

## Tests

Add:

* `shouldSkip` works with `retries: 0` (should continue retrying until a non-skipped error or time budget). This currently fails with your ordering.
* Async `shouldSkip` is awaited (once implemented).
* Backoff behavior around skipped errors (pick a deterministic minTimeout and assert elapsed time growth—or explicitly assert *no* growth if you choose that model).
* `shouldSkip` does not bypass `AbortError` / non-network `TypeError` (these should still short-circuit). ([[GitHub](https://github.com/sindresorhus/p-retry)][1])
* `onFailedAttempt` still fires for skipped errors and receives consistent `attemptNumber`/`retriesLeft`.

## Minor

* Default: keep `options.shouldSkip ??= () => false;` but move the call earlier as above.
* Keep internal symmetry: `onAttemptFailure(context, …)` change is fine, but also consider building the context once with the **effective** attempt number.

### Summary

Good feature; current implementation miscounts at the edges and is inconsistent with the v7 context-based API. Fix the ordering, expose `context`, allow async, clarify docs, and extend tests. After that, ship.

[1]: https://github.com/sindresorhus/p-retry "GitHub - sindresorhus/p-retry: Retry a promise-returning or async function"

timkelty committed:
Pass RetryContext

timkelty committed:
Clarify retry logic

timkelty committed:
Use retriesLeft

timkelty committed:
WS

timkelty committed:
Addressing AI PR review

timkelty committed:
xo

timkelty committed:
Normalized error in context test

timkelty commented:
I've addressed the 🤖's concerns and added tests for everything.

The tweaked design now uses the `RetryContext` more broadly and consistently to do most of the normalization and calculation, which simplifies other parts of the code, and makes things easier to understand.

timkelty committed:
Don't need skip examples here

timkelty committed:
Don't need skip examples here

timkelty committed:
Words

timkelty committed:
Don't need skip examples here

timkelty committed:
Trim examples

timkelty committed:
Restore changes

sindresorhus commented on index.d.ts:
```diff
@@ -14,6 +14,8 @@ export type RetryContext = {
 	readonly attemptNumber: number;
 	readonly retriesLeft: number;
+	readonly skippedRetries: number;
+	readonly skip: boolean;
```
`startTime` and `maxRetryTime` missing.

sindresorhus commented:
Nice work!

---

Another AI review. Mostly just minor stuff:

---

* **Doc gaps:** Update the **`onFailedAttempt(context)`** section to mention the new `context.skip` and `context.skippedRetries`, and that `onFailedAttempt` still runs for skipped errors. Right now only the new `shouldSkip` section explains skips; `onFailedAttempt` consumers won’t know these fields exist.

* **Wrong error forwarded when `shouldSkip` throws:** If `shouldSkip` throws, you call `onAttemptFailure` with the *original* operation error instead of the thrown one, which is misleading for `onFailedAttempt`. Build a new context with the thrown error:
  `index.js` (`createRetryContext` catch)

```js
} catch (thrown) {
const thrownError = normalizeError(thrown);
await onAttemptFailure(Object.freeze({...context, error: thrownError}), options);
throw thrownError;
}
```

* **Double “TypeError (non-network)” filtering:** You normalize in `normalizeError`, then check again in `onAttemptFailure`. Pick one place (prefer `normalizeError`) to avoid duplication. Either remove the check in `onAttemptFailure` or stop doing it in `normalizeError`.

* **Edge-case docs:** Clarify in README that `skip` still respects `maxRetryTime` (you short-circuit delay but still abort when time is exhausted). The code does this; the docs should say it.

* **Minor naming nit (optional):** `shouldSkip` is fine, but if you want symmetry with `shouldRetry`, consider explicitly stating “skips don’t consume budget and don’t advance backoff; attempt numbering remains 1-based for non-skipped attempts.” The tests imply this; spell it out.

timkelty committed:
Fix type def

timkelty committed:
Docs clarification

timkelty committed:
PR adjustments

timkelty committed:
Docs

timkelty committed:
Docs

timkelty committed:
Fix

timkelty committed:
Fix

timkelty committed:
clarify

timkelty committed:
Sync type def

timkelty committed:
Docs update

timkelty commented:
Reworked things a bit to simplify and resolve remaining notes:

- `onAttemptFailure` now returns a boolean, determining if it should expend a retry or not
- Simplified context to only add `retriesUsed`
- The callbacks all now note what happens if they throw
- I kept it as `shouldSkip`, but would be fine with something more explicit like `shouldUseRetry`, `shouldConsumeRetry`, `shouldExpendRetry`, `shouldIncrementRetry`

timkelty commented:
@sindresorhus FWIW, while working on this, I realized how somewhat hairy and ambiguous the concepts and language around "retry", "skip", etc. can be (even the existing `shouldSkip` [wasn't clear to me](https://github.com/sindresorhus/p-retry/pull/93) without diving into the code)

I'm wondering if a better/more flexible solution might be to let the consumer manipulate the `RetryContext`, either via a `onCreateRetryContext` option, or maybe by having `onFailedAttempt` optionally return a `RetryContext`.

That way there is no assumptions as to what "skip" means, and the consumer can have full control over the context that gets passed to the other callbacks.

Even so, I still think this PR has some worthwhile improvements (expanding the use of `RetryContext`). If that approach sounds worthwhile/better, let me know and I can make a separate PR.

sindresorhus commented:
I think we should go with the `shouldConsumeRetry` name.

sindresorhus commented:
Docs: Call order + TypeError / AbortError interaction
raedme adds `shouldSkip` but doesn’t explain the order. With current behavior:

* `AbortError` short-circuits **before** any callbacks (`onFailedAttempt`, `shouldSkip`, etc).
* `shouldSkip` runs **before** the non-network `TypeError` guard, meaning it can “rescue” a `TypeError` by skipping it; `shouldRetry` runs after.

Spell that out in both `shouldRetry` and `shouldSkip` docs. Keep the guarantee that normal `TypeError`s don’t retry unless they’re network-related.

---

**Zero-delay: allow `minTimeout: 0`**
`calculateDelay` currently does `Math.max(options.minTimeout, 1)`. Either allow true `0ms` delay or document the 1ms floor. Allowing 0 seems better.

---

**Nit: use monotonic time for `maxRetryTime`**
Use `performance.now()` deltas instead of `Date.now()` to avoid issues from wall-clock jumps (like sleep/wake).

---

### Tests to add

* **TypeError skip override**
  Throw a non-network `TypeError`, `shouldSkip` returns `true`, then throw a regular `Error`. Assert it continued and backoff stayed the same.

* **`minTimeout: 0`**
  With `randomize: false`, assert delays start at `[0, …]`.

* **AbortError short-circuit**
  Assert no callbacks fire and the error bubbles right away.

sindresorhus commented:
There could be a more powerful version, but I prefer to ship this one first and then we can see whether it's needed. Should be an issue first so we can discuss potential API. One idea I had was something like this:

```js
import pRetry from 'p-retry';

const result = await pRetry(run, {
	onFailedAttempt(context) {
		if (context.error instanceof RateLimitError) {
			return pRetry.skipWithDelay(context.error.retryAfter * 1000);
		}

		if (context.error.statusCode >= 500) {
			return pRetry.retryWithDelay(2000); // Consume retry, custom delay
		}

		if (context.error.statusCode === 404) {
			return pRetry.abort(); // Stop immediately
		}

		// No return = default behavior
	}
});
```

timkelty committed:
`shouldSkip` renamed to `shouldConsumeRetry`

timkelty committed:
Allow minTimeout: 0

timkelty committed:
Use performance.now

timkelty committed:
Clarify docs callback

timkelty committed:
Sync type def

timkelty committed:
xo

timkelty committed:
AbortError test

timkelty committed:
Add backoff/consume test

timkelty commented:
- [x]  `shouldSkip` renamed to `shouldConsumeRetry`
- [x] Docs: Call order + TypeError / AbortError interaction
- [x] use monotonic time for maxRetryTime
- [x] Tests to add
    - [x] TypeError skip override
    - [x] `minTimeout: 0`
    - [x] AbortError short-circuit

Also fixed fix a bug where `shouldConsumeRetry() => false` wouldn't call `shouldRetry`, and would retry forever if `retries: 0`.

timkelty committed:
Fix logic bug

sindresorhus commented:
This looks good now. Thanks for your hard work on this 🙏
